### DISCUSSION TITLE
Moodle LMS 1.3 integration roster issues

### DISCUSSION THREAD

**Discussion by Syxton** [community/community#40468] [GitHub Education]
I have a faculty member that is wanting to use Github Classrooms integration with Moodle.
Our IT department followed the instructions and registered our Moodle instance and created the external Tool.
When the faculty member tested it, his roster didn't import all the correct students.
I then tested it myself in a different course, and the only options seemed to be to setup a classroom would overwrite his classroom (even though we were using two different Moodle courses).  I seemed to be able to get my course roster to import, but when he tried again from his Moodle course, he only got my courses roster.  After I deleted my organization and classroom he was able to get back to the incorrect roster he was getting previously.

**Comment by ossanna16** [comment]
Hi there @Syxton and welcome to our community! Thank you for asking a great question 🙂
While you are waiting for a reply to your question, here are a few things you can do:

To get started, introduce yourself in our [official introduction thread](https://github.com/orgs/community/discussions/23204)
To connect with other new users, check out this Discussion to [Meet Your GitHub Community Team](https://github.com/orgs/community/discussions/28776), further your GitHub knowledge by working through our [GitHub Skill 🆙 exercises](https://github.com/orgs/community/discussions/32121), and participate in our [weekly Fun Friday chats](https://github.com/orgs/community/discussions/30912)

**Comment by regisin** [comment]
I just wanted to bump this up. I have similar issues. I'm the instructor, not Moodle admin.

After the Moodle instance was registered, I tried importing the roster. If I sync names or emails, GH Classroom doesn't seem to find any student enrolled. If I try with IDs, it imports like 5 records, none of which is a valid student ID, just some strange numbers I can't identify what they are (also, the classroom has 25 students enrolled).

I'm happy to provide more details, or ask the IT admin to get involved.

**Comment by jyounggo** [comment]
Same issue as @regisin.
When I try to import roster through "Names" or "Emails", I got "No valid identifiers in provided Roster Entries"
![image](https://user-images.githubusercontent.com/17755569/211945663-8f24b70a-f962-4e3d-926c-449a7614d712.png)

"User IDs" seems working but it looks like it is just user_id from Moodle context.

**Comment by samus-aran** [comment]
Hi everyone 👋 

We've raised this with the internal team and they have identified missing gaps in the documentation and will update it accordingly. I'll post an update to this thread when that change has been published 👍

**Reply by jyounggo** [reply]
@samus-aran  Do you have any update on this? Still the team is working on this?

**Comment by Syxton** [comment]
This integration is still broken.  Faculty are getting course rosters that are completely incorrect.  It also seems that they can only get 1 roster at a time, even though they are creating separate Github Classrooms.  Like an organization can only have 1 roster at a time.

**Reply by nvidot** [reply]
Regarding the 1 roster limit, it may be because the procedure describing how to create the link between LMS and Classroom is misleading. What worked for me is to add an "External Tool" activity, then add a new "Preconfigured External Tool" from there. Once done, when we access the external tool, we are invited to register the lms. We can do so using the informations from "Manage Tools" in "Site Administration > Plugins" in Moodle.  Repeat this operation in each moodle course that will serve as a source for a classroom roster.
Regarding the "No valid entries", setting "Share launcher's name with tool" to "Always", seems to have fixed that for retrieving "Names" when syncing.

**Reply by elearningservicesgithub** [reply]
I was able to successfully integrate a Github classroom with my Moodle course and sync the names. But, when I test it in another course it imports the same roster. As you explained above, when I try to create an external tool and connect the Github classroom for the first time it asks to register with the LMS. At that point the Github classroom and the Moodle course establish a connection. The workaround you suggest does not work for me as it does not allow me to register the LMS as it says the 'LMS is already registered'. Any advice would be helpful. Thanks

**Reply by elearningservicesgithub** [reply]
I was able to import rosters by using the unique client ids generated for each course instance.

**Comment by regisin** [comment]
Another issue, I can't update the roster. Since the integration is not working, I added the roster manually. But I have a student that enrolled after I did it, and now there is no way I can add the student so he can link his GH user to the classroom.

**Reply by mkwapisz** [reply]
We are facing the same problem. The roaster once read from a moodle course cannot be updated. The synchronization process provides information that new students have been found, but they are not displayed.

**Reply by mkwapisz** [reply]
![wikamp_Sltisync](https://github.com/user-attachments/assets/3d4afa69-ee6e-443d-b9e3-f4da21841da0)
I am not a system administrator, only a simple teacher. Test scenario:

- create a new roaster by Import from moodle (there are 2)
- add new students to moodle course - now there are 4
- from the roaster choose to synchronise from moodle
- only 2 "old" students are visible, regardless github states 4 elements have been synchronized, see picture attached.
-

**Reply by elms-lcc** [reply]
I'm also facing this issue. I'd like to request at least a manual method to add student so I could work around this issue.

**Comment by barhydtk** [comment]
Hi folks, sorry to jump in late here, but I just tried to use the LTI with Moodle and seem to have the same issues listed here.

I've been trying to look through the documentation from GitHub and am not really understanding what I'm seeing. Do they have direct support that you could check with them? It appears the LTI is connecting correctly, and past that, we don't have much insight into what is going on with the LTI provider side.

Can anyone point me to any resolution or guidance?
Thank you,
Kevin

**Reply by Syxton** [reply]
@barhydtk  I'm the originator of this issue, and it has been 4 months and I have had no response or fix.

**Reply by barhydtk** [reply]
Thank you for the reply.
Have a good weekend,
Kevin

**Comment by nvidot** [comment]
Hi everyone,

The procedure given in GitHub documentation result to connect a _single_ "Moodle course participants" _list_ to a _single_ "GitHub classroom". So you have to create an external tool for each _distinct_ Moodle course participants _list_ you'll want connected; if you were to create only one external tool then you will enconter the one roster problem describe in [Syxton message above](https://github.com/orgs/community/discussions/40468#discussioncomment-4756576).

_From what I understand_, the given procedure should however result in configuring a "preconfigured external tool" which will then serve in each Moodle course to generate the _actual_ external tool (one with the course_id needed by GH Classroom to retrieve the course participants). For some reason, in recent Moodle, one cannot use a preconfigured external tool that way because at creation time the external tool configuration does not offer an option to do so (it used to do so in a previous version of Moodle)... and trying to use the "preconfigured tool" directly as an activity result in an error beacuse of "missing context" (probably course_id is missing).

However you can create an external tool directly at the course level (given that you have the adequate permission). For that, go in "Plus", choose "LTI external tool" then push "Add tool" button and configure the tool according to GH documentation. For previous versions of Moodle, as  i remember, you can add an activity and choose to configure an external tool from there. With recent version You only need to configure the tool, no need to add it as an activity on your page.

Register the tool with GH. You will need to have administrative rights to go to "Administration" > "Plugins" > "External Tool" > "Manage tools" to retrieve the informations needed; But if your Moodle admin make the various URL (which are constants) available you can retrieve the "Client ID" from the external tool in your course and procede to the registration !

If all was ok but you cannot retrieve the moodle participants when creating the roster it might be that authentication headers are lost (error 401 when retrieving membership in web server log) between Apache and PHP.

The fix is to set this (in the httpd.conf file or in the vhosts file) :

    SetEnvIf Authorization "(.*)" HTTP_AUTHORIZATION=$1
    SetEnvIf Content-Type "(.*)" HTTP_CONTENT_TYPE=$1
    SetEnvIf Accept "(.*)" HTTP_ACCEPT=$1

(source: https://moodle.org/mod/forum/discuss.php?d=389429)

**Comment by creamcollar1122** [comment]
facing issue with integrating moodle with class room Error connecting to GitHub Classroom
Classroom may be misconfigured on your LMS. Contact your LMS administrator.

Error: Missing required information: Context Membership URL, followed all the steps in documents still same issue

**Comment by finnominimality** [comment]
Hi,

I'm encountering the same sync issue as mentioned [above](https://github.com/orgs/community/discussions/40468#discussioncomment-4758013). Any updates on this?

Additionally, I can't import the roster using the "User IDs". Names and Email works, but for anonymization reasons I'd need to import the User ID. All enrolled students have an "ID number" assigned in their Moodle profile. Any advice here? 

Many thanks!
